### DISCUSSION TITLE
Best way to create a sitemap for both static and dynamic pages

### DISCUSSION THREAD

**Discussion by choutkamartin** [vercel/next.js#27558] [Help]
Hello,
I want to open a discussion about sitemap creation in Next.js. I have an app that has static and dynamic pages (blog articles).

### Description
I've created a page sitemap.xml.js that doesn't render anything but sends the sitemap response instead. I use globby for scanning the static pages, and I fetch my database for the blog articles. The code is here:

```js
import globby from "globby";
import absoluteUrl from "next-absolute-url";

const Sitemap = () => {};

export const getServerSideProps = async ({ req, res }) => {
  const { protocol, host } = absoluteUrl(req, "localhost:3000");
  const data = await fetch(`${protocol}//${host}/api/articles`);
  const dynamicPages = await data.json();
  const dynamicSlug = dynamicPages.articles.map(
    (dynamicPage) => `blog/${dynamicPage.slug}`
  );
  const staticPages = await globby([
    "./",
    "!node_modules",
    "!public/",
    "!components/",
    "!api/",
  ]);
  const pages = [...staticPages, ...dynamicSlug];
  const sitemap = `<?xml version="1.0" encoding="UTF-8"?>
  <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
    xmlns:xhtml="http://www.w3.org/1999/xhtml">
      ${pages
        .map((url) => {
          const path = url
            .replace("pages/", "")
            .replace(".jsx", "")
            .replace(".js", "")
            .replace(".mdx", "")
            .replace("index", "");
          return `
            <url>
              <loc>${protocol}//${host}/${path}</loc>
              <xhtml:link
               rel="alternate"
               hreflang="en"
               href='${protocol}//${host}/en/${path}'/>
            </url>
          `;
        })
        .join("")}
    </urlset>
  `;
  res.setHeader("Content-Type", "text/xml");
  res.write(sitemap);
  res.end();

  return {
    props: {},
  };
};

export default Sitemap;

```

### The problem
When I build the app and run it on localhost, everything works correctly. But when I deploy to Vercel, the static pages aren't scanned by globby. I think that happens because Vercel builds the file/folder structure a little differently.

### My solution?
I thought about creating a sitemap for static files at a build time and a sitemap as a page for dynamic pages. Then use a sitemap index file to let search engines know there are two sitemaps.

This way, I can easily control the "last modified" date for both static and dynamic pages.

PS: I tried some npm packages and checked out some articles on the internet, but I was never satisfied.

### So, what is the best way to create sitemaps in Next.js in your opinion?
I want to ask if you have an any better idea of how to make a static/dynamic sitemap.

Thanks, I hope you have a great day!

**Comment by zzimickQ** [comment]
I've tried the exact same methods you did, but ended up not meeting my goal. I'm now checking the nextjs build process to create webpack plugin to tap into the build process and generate a sitemap on the right build step. If you're open we can look at it together.

**Reply by choutkamartin** [reply]
I will take a look at some other options how I could build sitemap easily, and I will also take a look at the build process, and get in touch with you later if I would find something that would work. I don't understand the build process quite yet, so it will probably take me some time.

**Comment by bcavs** [comment]
did you find a good solution for this? I am having the same problem with the static files not being found in the Vercel deployments

**Reply by choutkamartin** [reply]
I ended up using next-sitemap package which works well, where I server-side render the sitemap.

On some sites I dropped SSG completely as it was hard maintain with a lot of pages (serverless revalidate function started to time-out as revalidating 11 sites took longer than 10 seconds).

Now I run both CMS and Next.js on one VPS as Docker containers. If I need global CDN I use Cloudflare and I don't need to worry about Vercel limits and I have more freedom with everything
